hmmlearn/hmm.py

```diff
@@ -520,134 +520,221 @@ class GMMHMM(_BaseHMM):
+
+        if 'w' in self.init_params or not hasattr(self, 'weights_'):
+            self.weights_ = (np.ones((self.n_components, self.n_mix)) /
+                             (np.ones((self.n_components, 1)) * self.n_mix))
+
+        if 'm' in self.init_params or not hasattr(self, 'means_'):
+            self.means_ = np.zeros((self.n_components, self.n_mix,
+                                    self.n_features))
+            for i, kmeans in enumerate(kmeanses):
+                self.means_[i] = kmeans.cluster_centers_
+
+        if 'c' in self.init_params or not hasattr(self, 'covars_'):
+            cv = np.cov(X.T)
+            if not cv.shape:
+                cv.shape = (1, 1)
+
+            if self.covariance_type == 'tied':
+                self.covars_ = np.zeros((self.n_components,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'full':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'diag':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features))
+                self.covars_[:] = np.diag(cv)
+            elif self.covariance_type == 'spherical':
+                self.covars_ = np.zeros((self.n_components, self.n_mix))
+                self.covars_[:] = cv.mean()
+
+    def _check(self):
+        super(GMMHMM, self)._check()
 
     def _generate_sample_from_state(self, state, random_state=None):
-        return self.gmms_[state].sample(1, random_state=random_state).flatten()
+        # TODO: actually use random_state
+        if random_state is None:
+            random_state = self.random_state
+
+        cur_means = self.means_[state]
+        cur_covs = self.covars_[state]
+        cur_weights = self.weights_[state]
+
+        i_gauss = np.random.multinomial(1, cur_weights)[0][0]
+        mean = cur_means[i_gauss]
+        if self.covariance_type == 'tied':
+            cov = cur_covs
+        else:
+            cov = cur_covs[i_gauss]
+
+        return sample_gaussian(mean, cov, self.covariance_type,
+                               random_state=random_state)
+
+    def _compute_log_likelihood(self, X):
+        n_samples, _ = X.shape
+        res = np.zeros((n_samples, self.n_components))
+
+        for i in range(self.n_components):
+            cur_means = self.means_[i]
+            cur_covs = self.covars_[i]
+            if self.covariance_type == 'spherical':
+                cur_covs = cur_covs[:, np.newaxis]
+            cur_weights = self.weights_[i]
+
+            densities = np.exp(log_multivariate_normal_density(
+                X, cur_means, cur_covs, self.covariance_type
+            ))
+            res[:, i] = np.log(np.sum(cur_weights * densities, axis=1))
```

Comment:
This is usually done using [`logsumexp`](http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.misc.logsumexp.html) from `scipy.misc`:

```
res[:, i] = logsumexp(log_multivariate_normal_density(...) + cur_weights, axis=1)
```